src/application-manager.coffee

```diff
@@ -0,0 +1,1047 @@
+		Promise.join(
+			@services.getAllByAppId(appId)
+			@networks.getAllByAppId(appId)
+			@volumes.getAllByAppId(appId)
+			(services, networks, volumes) =>
+				return @_buildApps(services, networks, volumes)[appId]
+		)
+
+	getTargetApp: (appId) =>
+		@db.models('app').where({ appId }).select()
+		.then ([ app ]) =>
+			if !app?
+				return
+			@normaliseAndExtendAppFromDB(app)
+
+	# Compares current and target services and returns a list of service pairs to be updated/removed/installed.
+	# The returned list is an array of objects where the "current" and "target" properties define the update pair, and either can be null
+	# (in the case of an install or removal).
+	compareServicesForUpdate: (currentServices, targetServices) =>
+		removePairs = []
+		installPairs = []
+		updatePairs = []
+		targetServiceIds = _.map(targetServices, 'serviceId')
+		currentServiceIds = _.uniq(_.map(currentServices, 'serviceId'))
+
+		toBeRemoved = _.difference(currentServiceIds, targetServiceIds)
+		for serviceId in toBeRemoved
+			servicesToRemove = _.filter(currentServices, (s) -> s.serviceId == serviceId)
+			for service in servicesToRemove
+				removePairs.push({
+					current: service
+					target: null
+					serviceId
+				})
+
+		toBeInstalled = _.difference(targetServiceIds, currentServiceIds)
+		for serviceId in toBeInstalled
+			serviceToInstall = _.find(targetServices, (s) -> s.serviceId == serviceId)
+			if serviceToInstall?
+				installPairs.push({
+					current: null
+					target: serviceToInstall
+					serviceId
+				})
+
+		toBeMaybeUpdated = _.intersection(targetServiceIds, currentServiceIds)
+		currentServicesPerId = {}
+		targetServicesPerId = _.keyBy(targetServices, 'serviceId')
+		for serviceId in toBeMaybeUpdated
+			currentServiceContainers = _.filter currentServices, (service) ->
+				return service.serviceId == serviceId
+			if currentServiceContainers.length > 1
+				currentServicesPerId[serviceId] = _.maxBy(currentServiceContainers, 'createdAt')
+				# All but the latest container for this service are spurious and should be removed
+				for service in _.without(currentServiceContainers, currentServicesPerId[serviceId])
+					removePairs.push({
+						current: service
+						target: null
+						serviceId
+					})
+			else
+				currentServicesPerId[serviceId] = currentServiceContainers[0]
+
+		# Returns true if a service matches its target except it should be running and it is not, but we've
+		# already started it before. In this case it means it just exited so we don't want to start it again.
+		alreadyStarted = (serviceId) =>
+			currentServicesPerId[serviceId].isEqualExceptForRunningState(targetServicesPerId[serviceId]) and
+				targetServicesPerId[serviceId].running  and
+				@_containerStarted[currentServicesPerId[serviceId].containerId]
+
+		needUpdate = _.filter toBeMaybeUpdated, (serviceId) ->
+			return !currentServicesPerId[serviceId].isEqual(targetServicesPerId[serviceId]) and !alreadyStarted(serviceId)
```

Comment:
NIT: Unnecessary return (normally I wouldn't say anything but this line is pretty long as it is).